CryptoPkg/Library/BaseCryptLibOnOneCrypto/OneCryptoLib.c

```diff
@@ -0,0 +1,6344 @@
+    ASSERT (FALSE);
+    return NULL;
+  }
+
+  if (!ValidateCryptoVersion (gCryptoProtocol, MinMajor, MinMinor)) {
+    return NULL;
+  }
+
+  return gCryptoProtocol;
+}
+
+/**
+  Internal worker function that prints a debug message and asserts if a crypto
+  service is not available.  This should never occur because library instances
+  have a dependency expression for the for the One Crypto Protocol so
+  a module that uses these library instances are not dispatched until the One
+  Crypto Protocol is available.  The only case that this function handles is
+  if the One Crypto Protocol installed is NULL or a function pointer in
+  the One Protocol is NULL.
+
+  @param[in]  FunctionName  Null-terminated ASCII string that is the name of an
+                            One Crypto service.
+
+**/
+static
+VOID
+CryptoServiceNotAvailable (
+  IN CONST CHAR8  *FunctionName
+  )
+{
+  DEBUG ((DEBUG_ERROR, "[%a] Function %a is not available\n", gEfiCallerBaseName, FunctionName));
+  ASSERT_EFI_ERROR (EFI_UNSUPPORTED);
+}
+
+/**
+  Internal worker function to validate a crypto function pointer.
+
+  This function checks if the specified function pointer is available.
+
+  @param[in]  FunctionPtr   Pointer to check.
+  @param[in]  FunctionName  Name of the function (for debug output).
+
+  @retval TRUE   Function pointer is available.
+  @retval FALSE  Function pointer is NULL.
+**/
+static
+BOOLEAN
+ValidateCryptoFunction (
+  IN CONST VOID   *FunctionPtr,
+  IN CONST CHAR8  *FunctionName
+  )
+{
+  if (FunctionPtr == NULL) {
+    CryptoServiceNotAvailable (FunctionName);
+    return FALSE;
+  }
+
+  return TRUE;
+}
+
+/**
+  A macro used to call a non-void service in an One Crypto Protocol.
+  If the protocol is NULL or the service in the protocol is NULL, then a debug
+  message and assert is generated and an appropriate return value is returned.
+
+  This macro allows specifying minimum version requirements per function call,
+  enabling backward compatibility with older crypto binaries.
+
+  @param  Function          Name of the One Crypto Protocol service to call.
+  @param  Args              The argument list to pass to Function.
+  @param  ErrorReturnValue  The value to return if the protocol is NULL or the
+                            service in the protocol is NULL.
+  @param  MinMajor          Minimum required major version for this function.
+  @param  MinMinor          Minimum required minor version for this function.
+
+**/
+#define CALL_CRYPTO_SERVICE(Function, Args, ErrorReturnValue, MinMajor, MinMinor)  \
+  do {                                                                 \
+    ONE_CRYPTO_PROTOCOL  *CryptoServices;                            \
+                                                                       \
+    CryptoServices = GetAndValidateCryptoProtocol (#Function, (MinMajor), (MinMinor)); \
+    if ((CryptoServices == NULL) || !ValidateCryptoFunction ((VOID *)CryptoServices->Function, #Function)) { \

```

Comment:
Casting a function pointer to `VOID *` is not portable will be a problem in some compilers, you will need to do something else.